File: admin/src/main/java/com/beat/admin/promotion/api/AdminPromotionController.java
package com.beat.admin.promotion.api;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.beat.admin.api.response.AdminSuccessCode;
import com.beat.admin.promotion.application.dto.request.CarouselHandleRequest;
import com.beat.admin.promotion.application.dto.response.BannerPresignedUrlFindResponse;
import com.beat.admin.promotion.application.dto.response.CarouselFindAllResponse;
import com.beat.admin.promotion.application.dto.response.CarouselHandleAllResponse;
import com.beat.admin.promotion.application.dto.response.CarouselPresignedUrlFindAllResponse;
import com.beat.admin.promotion.facade.AdminPromotionFacade;
import com.beat.gateway.security.servlet.CurrentMember;
import com.beat.global.support.response.SuccessResponse;

import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/admin")
@RequiredArgsConstructor
public class AdminPromotionController implements AdminPromotionApi {

	private final AdminPromotionFacade adminPromotionFacade;

	@Override
	@GetMapping("/carousels/presigned-url")
	public ResponseEntity<SuccessResponse<CarouselPresignedUrlFindAllResponse>> createAllCarouselPresignedUrls(
		@CurrentMember Long memberId, @RequestParam List<String> carouselImages) {
		CarouselPresignedUrlFindAllResponse response = adminPromotionFacade.checkMemberAndIssueAllPresignedUrlsForCarousel(
			memberId, carouselImages);
		return ResponseEntity.ok(SuccessResponse.of(AdminSuccessCode.CAROUSEL_PRESIGNED_URL_ISSUED, response));
	}

	@Override
	@GetMapping("/banner/presigned-url")
	public ResponseEntity<SuccessResponse<BannerPresignedUrlFindResponse>> createBannerPresignedUrl(
		@CurrentMember Long memberId, @RequestParam String bannerImage) {
		BannerPresignedUrlFindResponse response = adminPromotionFacade.checkMemberAndIssuePresignedUrlForBanner(memberId,
			bannerImage);
		return ResponseEntity.status(HttpStatus.OK)
			.body(SuccessResponse.of(AdminSuccessCode.BANNER_PRESIGNED_URL_ISSUED, response));
	}

	@Override
	@GetMapping("/carousels")
	public ResponseEntity<SuccessResponse<CarouselFindAllResponse>> readAllCarouselImages(
		@CurrentMember Long memberId) {
		CarouselFindAllResponse response = adminPromotionFacade.checkMemberAndFindAllPromotionsSortedByCarouselNumber(
			memberId);
		return ResponseEntity.status(HttpStatus.OK)
			.body(SuccessResponse.of(AdminSuccessCode.FETCH_ALL_CAROUSEL_PROMOTIONS_SUCCESS, response));
	}

	@Override
	@PutMapping("/carousels")
	public ResponseEntity<SuccessResponse<CarouselHandleAllResponse>> processCarouselImages(
		@CurrentMember Long memberId,
		@RequestBody CarouselHandleRequest request) {
		CarouselHandleAllResponse response = adminPromotionFacade.checkMemberAndProcessAllPromotionsSortedByCarouselNumber(
			memberId, request);
		return ResponseEntity.status(HttpStatus.OK)
			.body(SuccessResponse.of(AdminSuccessCode.UPDATE_ALL_CAROUSEL_PROMOTIONS_SUCCESS, response));
	}
}

File: admin/src/main/kotlin/com/beat/admin/promotion/api/AdminPromotionController.kt
package com.beat.admin.promotion.api

import com.beat.admin.promotion.api.request.CarouselHandleRequest
import com.beat.admin.promotion.api.response.BannerPresignedUrlFindResponse
import com.beat.admin.promotion.api.response.CarouselFindAllResponse
import com.beat.admin.promotion.api.response.CarouselHandleAllResponse
import com.beat.admin.promotion.api.response.CarouselPresignedUrlFindAllResponse
import com.beat.admin.promotion.api.response.PromotionSuccessCode
import com.beat.admin.promotion.facade.AdminPromotionFacade
import com.beat.gateway.CurrentMember
import com.beat.global.support.response.SuccessResponse
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/admin")
class AdminPromotionController(
    private val adminPromotionFacade: AdminPromotionFacade,
) : AdminPromotionApi {

    @GetMapping("/carousels/presigned-url")
    override fun createAllCarouselPresignedUrls(
        @CurrentMember memberId: Long,
        @RequestParam carouselImages: List<String>,
    ): ResponseEntity<SuccessResponse<CarouselPresignedUrlFindAllResponse>> {
        val response = adminPromotionFacade.checkMemberAndIssueAllPresignedUrlsForCarousel(memberId, carouselImages)
        return ResponseEntity.ok(SuccessResponse.of(PromotionSuccessCode.CAROUSEL_PRESIGNED_URL_ISSUED, response))
    }

    @GetMapping("/banner/presigned-url")
    override fun createBannerPresignedUrl(
        @CurrentMember memberId: Long,
        @RequestParam bannerImage: String,
    ): ResponseEntity<SuccessResponse<BannerPresignedUrlFindResponse>> {
        val response = adminPromotionFacade.checkMemberAndIssuePresignedUrlForBanner(memberId, bannerImage)
        return ResponseEntity.status(HttpStatus.OK)
            .body(SuccessResponse.of(PromotionSuccessCode.BANNER_PRESIGNED_URL_ISSUED, response))
    }

    @GetMapping("/carousels")
    override fun readAllCarouselImages(
        @CurrentMember memberId: Long,
    ): ResponseEntity<SuccessResponse<CarouselFindAllResponse>> {
        val response = adminPromotionFacade.checkMemberAndFindAllPromotionsSortedByCarouselNumber(memberId)
        return ResponseEntity.status(HttpStatus.OK)
            .body(SuccessResponse.of(PromotionSuccessCode.FETCH_ALL_CAROUSEL_PROMOTIONS_SUCCESS, response))
    }

    @PutMapping("/carousels")
    override fun processCarouselImages(
        @CurrentMember memberId: Long,
        @Valid @RequestBody request: CarouselHandleRequest,
    ): ResponseEntity<SuccessResponse<CarouselHandleAllResponse>> {
        val response = adminPromotionFacade.checkMemberAndProcessAllPromotionsSortedByCarouselNumber(memberId, request)
        return ResponseEntity.status(HttpStatus.OK)
            .body(SuccessResponse.of(PromotionSuccessCode.UPDATE_ALL_CAROUSEL_PROMOTIONS_SUCCESS, response))
    }
}